### PR TITLE
Refactor and Enhance Medical Condition Commands and Parsers

### LINKED ISSUE #149: Add SLAP to addMedCon and delMedCon
shift the guard conditions for medCon validity
to a separate isValid function in MedCon class

### PR DESCRIPTION
**Pull request by TechRaj**
- Add validation for medical condition fields to ensure consistency:
  - Limit character length to 30 characters.
  - Introduce proper error messages for invalid fields.
- Add utility functions to `ArgumentMultimap` to reduce duplication:
  - Moved `arePrefixesPresent` to `ArgumentMultimap`.
  - Introduced `requireAllPrefixesPresent` in `ParserUtil` for prefix validation.
- Improve code reuse across multiple parser classes:
  - Extracted common validation logic to `ParserUtil`.
  - Updated `AddMedConCommandParser` and `DelMedConCommandParser` to use the new utility functions.
- Modify success messages to provide more clarity.
- Update test cases:
  - Added test cases for new validation rules and `DelMedConCommand`.
  - Updated test cases to ensure proper failure for missing or incorrect fields.

Fixes #149

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -249,17 +251,20 @@ public static Set<Appointment> parseAppointments(Collection<String> appointments
...
     public static MedCon parseMedCon(String medConStr) throws ParseException {
-        requireNonNull(medConStr);
-        String trimmedMedCon = medConStr.trim();
-        return new MedCon(trimmedMedCon);
+        if (medConStr.isEmpty()) {
+            throw new ParseException("Medical condition " + MESSAGE_EMPTY_FIELD);
+        } else if (medConStr.length() > 30) {
+            throw new ParseException("Medical condition " + MESSAGE_CONSTRAINTS_LENGTH);
+        } else if (!medConStr.matches(MedCon.VALIDATION_REGEX)) {
+            throw new ParseException("Invalid medical condition format.");
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:266]: LGTM, no arrowhead

File: src/main/java/seedu/address/storage/JsonAdaptedMedCon.java [hunk 2]
```diff
@@ -39,6 +41,13 @@ public String getmedConName() {
      * @throws IllegalValueException if there were any data constraints violated in the adapted MedCon.
      */
     public MedCon toModelType() throws IllegalValueException {
+        if (medConName == null) {
+            throw new IllegalValueException(String.format("%s cannot be null", MedCon.class.getSimpleName()));
+        }
+        if (!MedCon.isValidMedCon(medConName)) {
+            throw new IllegalValueException(String.format("%s: %s", MedCon.class.getSimpleName(),
+                    MESSAGE_CONSTRAINTS_LENGTH));
+        }
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/storage/JsonAdaptedMedCon.java:50]: Good defensive coding here 👍🏻

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java [hunk 3]
```diff
@@ -75,4 +75,14 @@ public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseExcepti
...
+
+    /**
+     * Returns true if all of the specified prefixes are present and contain non-empty values.
+     *
+     * @param prefixes The prefixes to check.
+     * @return true if all prefixes are present, false otherwise.
+     */
+    public boolean arePrefixesPresent(Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> this.getValue(prefix).isPresent());
+    }
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:87]: Will all the other `add/delete` related commands use this? @TechRaj

**Review by yong-zhuo — APPROVED** [review]
Overall LGTM

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java (hunk 3 above)
Comment (TechRaj) [reply src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:87]: I basically shifted what @hweinian  did for allergy here so more CRUD functions can use it and comply with DRY

File: src/main/java/seedu/address/logic/Messages.java [hunk 4]
```diff
@@ -12,14 +12,15 @@
...
+    public static final String MESSAGE_EMPTY_FIELD = "field should not by empty";
     public static final String MESSAGE_UNKNOWN_COMMAND = "Unknown command";
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
     public static final String MESSAGE_PERSON_NRIC_NOT_FOUND = "The patient with the specified NRIC does not exist";
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d patients listed!";
     public static final String MESSAGE_DUPLICATE_FIELDS =
                 "Multiple values specified for the following single-valued field(s): ";
     public static final String MESSAGE_CONSTRAINTS_LENGTH =
-            "Medical condition length exceeds the limit of 45 characters";
+            "length should not exceed the limit of 30 characters";
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/Messages.java:23]: Good change, catering to other commands

File: src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java [hunk 5]
```diff
@@ -19,6 +19,7 @@
  * Parses user input for the {@link AddMedConCommand} and creates a new instance of it.
  */
 public class AddMedConCommandParser implements Parser<AddMedConCommand> {
+    private static final Logger logger = LogsCenter.getLogger(AddMedConCommandParser.class);
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java:22]: Good use of logger

File: src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java [hunk 6]
```diff
@@ -32,7 +33,8 @@ public AddMedConCommand parse(String args) throws ParseException {
         ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NRIC, PREFIX_MEDCON);
 
         // Check if NRIC is provided
-        if (!arePrefixesPresent(argMultimap, PREFIX_NRIC, PREFIX_MEDCON) || !argMultimap.getPreamble().isEmpty()) {
+        if (!argMultimap.arePrefixesPresent(PREFIX_NRIC, PREFIX_MEDCON) || !argMultimap.getPreamble().isEmpty()) {
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java:36]: Good defensive programming

File: src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java [hunk 7]
```diff
@@ -45,25 +47,13 @@ public AddMedConCommand parse(String args) throws ParseException {
...
-            if (medConStr.isEmpty()) {
-                throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
-                        AddMedConCommand.MESSAGE_USAGE));
-            } else if (medConStr.length() > 45) {
-                throw new ParseException(MESSAGE_CONSTRAINTS_LENGTH);
-            } else {
-                medCons.add(new MedCon(medConStr));
-            }
+            MedCon medCon = ParserUtil.parseMedCon(medConStr);
+            medCons.add(medCon);
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java:51]: Good SLAP

File: src/main/java/seedu/address/logic/Messages.java (hunk 4 above)
Comment (hweinian) [inline src/main/java/seedu/address/logic/Messages.java:23]: Good fix to match the UG

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 1 above)
Comment (PresleyChew) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:266]: Seconded!

File: src/test/java/seedu/address/storage/JsonAdaptedMedConTest.java [hunk 8]
```diff
@@ -18,6 +24,32 @@ public void toModelType_validMedCon_returnsMedCon() throws Exception {
...
+    }
+
+
+    @Test
+    public void toModelType_invalidMedConTooLong_throwsIllegalValueException() {
+        JsonAdaptedMedCon jsonAdaptedMedCon = new JsonAdaptedMedCon(INVALID_MEDCON_TOO_LONG);
+        assertThrows(IllegalValueException.class, "MedCon: " + MESSAGE_CONSTRAINTS_LENGTH, ()
+                -> jsonAdaptedMedCon.toModelType());
+    }
+
```
Comment (PresleyChew) [inline src/test/java/seedu/address/storage/JsonAdaptedMedConTest.java:52]: Comprehensive test cases!

**Review by PresleyChew — APPROVED** [review]
Overall LGTM!

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java (hunk 3 above)
Comment (hweinian) [reply src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:87]: I think this shift is ok. It aligns with DRY and reduces code duplication. How about the rest of the parts that are using it? Did you fix those commands?

**Review by hweinian — COMMENTED** [review]
overall LGTM. Need to check if there are other instances of arePrefixesPresent being used elsewhere when shifting it to another class.

**Pull request merged**